contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -380,60 +459,76 @@ contract LidoOracle is ILidoOracle, AragonApp {
-        uint256 mask = gatheredEpochData[_epochId].reportsBitMask;
-        uint256 popcnt = mask.popcnt();
-        if (popcnt < getQuorum())
-            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
+    function _getQuorumReport() internal view returns (bool isQuorum, Report memory modeReport) {
+        uint256 quorum = getQuorum();
 
-        assert(0 != popcnt && popcnt <= members.length);
+        // check most frequent cases: all reports are the same or no reports yet
+        if (gatheredReportKinds.length == 1) {
+            return (gatheredReportKinds[0].count >= quorum, gatheredReportKinds[0].report);
+        } else if (gatheredReportKinds.length == 0) {
+            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
+        }
 
-        // pack current gatheredEpochData mapping to uint256 array
-        uint256[] memory data = new uint256[](popcnt);
-        uint256 i = 0;
-        uint256 membersLength = members.length;
-        for (uint256 index = 0; index < membersLength; ++index) {
-            if (mask.getBit(index)) {
-                data[i++] = reportToUint256(gatheredEpochData[_epochId].reports[index]);
+        // if more than 2 kind of reports exist, choose the most frequent
+        uint256 maxi = 0;
+        for (uint256 i = 1; i < gatheredReportKinds.length; ++i) {
+            if (gatheredReportKinds[i].count > gatheredReportKinds[maxi].count) maxi = i;
+        }
+        for (i = 0; i < gatheredReportKinds.length; ++i) {
+            if (i != maxi && gatheredReportKinds[i].count == gatheredReportKinds[maxi].count) {
+                return (false, Report({beaconBalance: 0, beaconValidators: 0}));
             }
         }
-
-        assert(i == data.length);
-
-        // find mode value of this array
-        (bool isUnimodal, uint256 mode) = Algorithm.mode(data);
-        if (!isUnimodal)
-            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
-
-        // unpack Report struct from uint256
-        modeReport = uint256ToReport(mode);
-
-        return (true, modeReport);
+        return (gatheredReportKinds[maxi].count >= quorum, gatheredReportKinds[maxi].report);
     }
 
     /**
      * @dev Pushes the current data point if quorum is reached
      */
-    function _tryPush(uint256 _epochId) internal {
-        (bool isQuorum, Report memory modeReport) = _getQuorumReport(_epochId);
-        if (!isQuorum)
-            return;
+    function _tryPush(uint256 epochId) internal {
+        (bool isQuorum, Report memory modeReport) = _getQuorumReport();
+        if (!isQuorum) return;
 
         // data for this frame is collected, now this frame is completed, so
-        // minReportableEpochId should be changed to first epoch from next frame
+        // reportableEpochId should be changed to first epoch from the next frame
         BeaconSpec memory beaconSpec = _getBeaconSpec();
```

Comment:
It's already called _getBeaconSpec in the beginning of reportBeacon. Maybe we can reorganize code to cache this?

Reply:
Agree. Fix in https://github.com/lidofinance/lido-dao/pull/311.